package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -0,0 +1,688 @@
+    rvec       hbox_diag
+    rvec       mhbox_diag
+    real       max_cutoff2
+
+
+# Class to handle PBC calculations
+cdef class PBCBox(object):
+    cdef cPBCBox_t c_pbcbox
+    cdef rvec center
+    cdef rvec bbox_center
+    cdef bint is_triclinic
+
+    def __init__(self, real[:,::1] box):
+        self.update(box)
+
+    cdef void fast_update(self, real[:,::1] box) nogil:
+        cdef ns_int i, j, k, d, jc, kc, shift
+        cdef real d2old, d2new, d2new_c
+        cdef rvec trial, pos
+        cdef ns_int ii, jj ,kk
+        cdef ns_int *order = [0, -1, 1, -2, 2]
+        cdef bint use
+        cdef real min_hv2, min_ss, tmp
+
+        rvec_clear(self.center)
+        # Update matrix
+        self.is_triclinic = False
+        for i in range(DIM):
+            for j in range(DIM):
+                self.c_pbcbox.box[i][j] = box[i, j]
+                self.center[j] += 0.5 * box[i, j]
+
+                if i != j:
+                    if box[i, j] > EPSILON:
+                        self.is_triclinic = True
+            self.bbox_center[i] = 0.5 *  box[i, i]
+
+        # Update diagonals
+        for i in range(DIM):
+            self.c_pbcbox.fbox_diag[i]  =  box[i, i]
+            self.c_pbcbox.hbox_diag[i]  =  self.c_pbcbox.fbox_diag[i] * 0.5
+            self.c_pbcbox.mhbox_diag[i] = - self.c_pbcbox.hbox_diag[i]
+
+        # Update maximum cutoff
+
+        # Physical limitation of the cut-off
+        # by half the length of the shortest box vector.
+        min_hv2 = min(0.25 * rvec_norm2(&box[XX, XX]), 0.25 * rvec_norm2(&box[YY, XX]))
+        min_hv2 = min(min_hv2, 0.25 * rvec_norm2(&box[ZZ, XX]))
+
+        # Limitation to the smallest diagonal element due to optimizations:
+        # checking only linear combinations of single box-vectors (2 in x)
+        # in the grid search and pbc_dx is a lot faster
+        # than checking all possible combinations.
+        tmp = box[YY, YY]
+        if box[ZZ, YY] < 0:
+            tmp -= box[ZZ, YY]
+        else:
+            tmp += box[ZZ, YY]
+
+        min_ss = min(box[XX, XX], min(tmp, box[ZZ, ZZ]))
+        self.c_pbcbox.max_cutoff2 = min(min_hv2, min_ss * min_ss)
+
+    def update(self, real[:,::1] box):
+        if box.shape[0] != DIM or box.shape[1] != DIM:
+            raise ValueError("Box must be a {} x {} matrix. Got: {} x {})".format(
+                DIM, DIM, box.shape[0], box.shape[1]))
+        if (box[XX, XX] == 0) or (box[YY, YY] == 0) or (box[ZZ, ZZ] == 0):
+            raise ValueError("Box does not correspond to PBC=xyz")
+        self.fast_update(box)
+
+
+    cdef void fast_pbc_dx(self, rvec ref, rvec other, rvec dx) nogil:
+        cdef ns_int i, j
+        cdef rvec dx_start, trial
+
+        for i in range(DIM):
+            dx[i] = other[i] - ref[i]
+
+        for i in range (DIM-1, -1, -1):
+            while dx[i] > self.c_pbcbox.hbox_diag[i]:
+                for j in range (i, -1, -1):
+                    dx[j] -= self.c_pbcbox.box[i][j]

```

Comment:
@ayushsuhane Please have a close look at this loop. It does subtract the correct vector to land at point C in your example.

Reply:
Yes, I understand this loop. But I was wondering that even though the particles are in a brick shaped box, can we apply ``minimum_image_ortho`` here?  While the above loop will land at C, I think if we apply ``minimum_image`` from calc_distances.h, it will return distance corresponding to AD rather than AC.

Reply:
Even if the particles are put inside a brick-shaped box, the box itself is not brick-shaped so the loop is needed to use the correct box vectors and we can't just rely on `minimum_image_ortho`.